frame/tx-pause/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+	}
+
+	/// The set of calls that are explicitly paused.
+	#[pallet::storage]
+	pub type PausedCalls<T: Config> =
+		StorageMap<_, Blake2_128Concat, RuntimeCallNameOf<T>, (), OptionQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The call is paused.
+		IsPaused,
+
+		/// The call is unpaused.
+		IsUnpaused,
+
+		/// The call is whitelisted and cannot be paused.
+		Unpausable,
+
+		// The pallet or call does not exist in the runtime.
+		NotFound,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// This pallet, or a specific call is now paused.
+		CallPaused { full_name: RuntimeCallNameOf<T> },
+		/// This pallet, or a specific call is now unpaused.
+		CallUnpaused { full_name: RuntimeCallNameOf<T> },
+	}
+
+	/// Configure the initial state of this pallet in the genesis block.
+	#[pallet::genesis_config]
+	#[derive(DefaultNoBound)]
+	pub struct GenesisConfig<T: Config> {
+		/// The initially paused calls.
+		pub paused: Vec<RuntimeCallNameOf<T>>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			for call in &self.paused {
+				Pallet::<T>::ensure_can_pause(&call).expect("Genesis data is known good; qed");
+				PausedCalls::<T>::insert(&call, ());
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Pause a call.
+		///
+		/// Can only be called by [`Config::PauseOrigin`].
+		/// Emits an [`Event::CallPaused`] event on success.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::pause())]
+		pub fn pause(origin: OriginFor<T>, full_name: RuntimeCallNameOf<T>) -> DispatchResult {
+			T::PauseOrigin::ensure_origin(origin)?;
+
+			Self::do_pause(full_name).map_err(Into::into)
+		}
+
+		/// Un-pause a call.
+		///
+		/// Can only be called by [`Config::UnpauseOrigin`].
+		/// Emits an [`Event::CallUnpaused`] event on success.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::unpause())]
+		pub fn unpause(origin: OriginFor<T>, ident: RuntimeCallNameOf<T>) -> DispatchResult {
+			T::UnpauseOrigin::ensure_origin(origin)?;
+
+			Self::do_unpause(ident).map_err(Into::into)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub(crate) fn do_pause(ident: RuntimeCallNameOf<T>) -> Result<(), Error<T>> {
+		Self::ensure_can_pause(&ident)?;
+		PausedCalls::<T>::insert(&ident, ());
+		Self::deposit_event(Event::CallPaused { full_name: ident });
+
+		Ok(())
+	}
+
+	pub(crate) fn do_unpause(ident: RuntimeCallNameOf<T>) -> Result<(), Error<T>> {
+		Self::ensure_can_unpause(&ident)?;
+		PausedCalls::<T>::remove(&ident);
+		Self::deposit_event(Event::CallUnpaused { full_name: ident });
+
+		Ok(())
+	}
+
+	/// Return whether this call is paused.
+	pub fn is_paused(full_name: &RuntimeCallNameOf<T>) -> bool {
+		if T::WhitelistedCalls::contains(full_name) {
+			return false
+		}
+
+		<PausedCalls<T>>::contains_key(full_name)
+	}
+
+	/// Same as [`Self::is_paused`] but for non-MEL-bound inputs.
```

Comment:
it took me a sec to figure out what MEL meant, could un-abbreviate it for clarity
```suggestion
	/// Same as [`Self::is_paused`] but for inputs unbound by max-encoded-len.
```